### PR TITLE
FOGL-1159 Implement Management API interface to configuration management

### PR DESCRIPTION
**Pull request by praveen-garg**
Test result:
```
tests/unit/python/foglamp/common/microservice_management_client/
test_microservice_management_client.py ...................................... 
==================================================== 929 passed, 12 skipped in 43.99 seconds ====================================================

```
Code Coverage:
```
Coverage for python/foglamp/common/microservice_management_client/
microservice_management_client.py : 100%
```

### CODE REVIEW

File: python/foglamp/services/core/server.py [hunk 1]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
     @classmethod
     async def change(cls, request):
         pass
+
+    @classmethod
+    async def get_configuration_categories(cls, request):
+        return conf_api.get_categories(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

File: python/foglamp/services/core/server.py [hunk 2]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
...
+
+    @classmethod
+    async def create_configuration_category(cls, request):
+        return conf_api.create_category(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

File: python/foglamp/services/core/server.py [hunk 3]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
...
+
+    @classmethod
+    async def get_configuration_category(cls, request):
+        return conf_api.get_category(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

File: python/foglamp/services/core/server.py [hunk 4]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
...
+
+    @classmethod
+    async def get_configuration_item(cls, request):
+        return conf_api.get_category_item(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

File: python/foglamp/services/core/server.py [hunk 5]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
...
+
+    @classmethod
+    async def update_configuration_item(cls, request):
+        return conf_api.set_configuration_item(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

File: python/foglamp/services/core/server.py [hunk 6]
```diff
@@ -998,3 +1000,27 @@ def stop(cls, loop=None):
...
+
+    @classmethod
+    async def delete_configuration_item(cls, request):
+        return conf_api.delete_configuration_item_value(request)
```
Comment (amarendra-dianomic) [inline python/foglamp/services/core/server.py]: returning a coroutine instead of response?

**Comment by amarendra-dianomic** [comment]
This is a crucial PR, so please provide unit tests for at least new api methods for mm client.

**Comment by praveen-garg** [comment]
@amarendra-dianomic keeping 1154 in mind, I'm running out of time. Please do the needful with FOGL-1183

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 7]
```diff
@@ -166,3 +170,115 @@ def get_services(self, service_name=None, service_type=None):
...
+            raise client_exceptions.MicroserviceManagementClientError(status=r.status, reason=r.reason)
+        if r.status in range(500, 600):
+            _logger.error("Server error code: %d, Reason: %s", r.status, r.reason)
+            raise client_exceptions.MicroserviceManagementClientError(status=r.status, reason=r.reason)
+        res = r.read().decode()
+        self._management_client_conn.close()
+        response = json.loads(res)
+        return response
+
+    def update_configuration_item(self, category_name, config_item, category_data):
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: thanks for fixing this!

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 8]
```diff
@@ -166,3 +170,115 @@ def get_services(self, service_name=None, service_type=None):
...
+        return response
+
+    def delete_configuration_item(self, category_name, config_item):
+        """
+
+        :param category_name:
+        :param config_item:
+        :return:
+        """
+        url = "/foglamp/service/category/{}/{}".format(category_name, config_item)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: let's keep the route URL consistent with Config Admin Api

Comment (amarendra-dianomic) [reply python/foglamp/common/microservice_management_client/microservice_management_client.py]: done

File: tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py [hunk 9]
```diff
@@ -267,3 +267,224 @@ def test_get_services_client_err(self, status_code):
...
                     ret_value = ms_mgt_client.get_services('foo', 'bar')
+
+    def test_get_configuration_category(self):
+        microservice_management_host = 'host1'
+        microservice_management_port = 1
+        ms_mgt_client = MicroserviceManagementClient(
+            microservice_management_host, microservice_management_port)
+        response_mock = MagicMock(type=HTTPResponse)
+        undecoded_data_mock = MagicMock()
+        response_mock.read.return_value = undecoded_data_mock
```
Comment (ashish-jabble) [inline tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py]: Better to make a fixture and use it , as above lines are repeated for every test.

File: tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py [hunk 10]
```diff
@@ -267,3 +267,224 @@ def test_get_services_client_err(self, status_code):
...
+        response_mock.status = status_code
+        with patch.object(HTTPConnection, 'request') as request_patch:
+            with patch.object(HTTPConnection, 'getresponse', return_value=response_mock) as response_patch:
+                with pytest.raises(client_exceptions.MicroserviceManagementClientError) as excinfo:
+                    ret_value = ms_mgt_client.get_configuration_category("SMNTR")
```
Comment (ashish-jabble) [inline tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py]: no assertion?

File: tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py [hunk 11]
```diff
@@ -267,3 +267,224 @@ def test_get_services_client_err(self, status_code):
...
+        ms_mgt_client = MicroserviceManagementClient(
+            microservice_management_host, microservice_management_port)
+        response_mock = MagicMock(type=HTTPResponse)
+        undecoded_data_mock = MagicMock()
+        response_mock.read.return_value = undecoded_data_mock
+        response_mock.status = status_code
+        with patch.object(HTTPConnection, 'request') as request_patch:
+            with patch.object(HTTPConnection, 'getresponse', return_value=response_mock) as response_patch:
+                with pytest.raises(client_exceptions.MicroserviceManagementClientError) as excinfo:
+                    ret_value = ms_mgt_client.get_configuration_item("SMNTR", "ping_timeout")
```
Comment (ashish-jabble) [inline tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py]: use only `ms_mgt_client.get_configuration_item("SMNTR", "ping_timeout")`

File: tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py [hunk 12]
```diff
@@ -267,3 +267,224 @@ def test_get_services_client_err(self, status_code):
...
+        ms_mgt_client = MicroserviceManagementClient(
+            microservice_management_host, microservice_management_port)
+        response_mock = MagicMock(type=HTTPResponse)
+        undecoded_data_mock = MagicMock()
+        response_mock.read.return_value = undecoded_data_mock
+        response_mock.status = status_code
+        with patch.object(HTTPConnection, 'request') as request_patch:
+            with patch.object(HTTPConnection, 'getresponse', return_value=response_mock) as response_patch:
+                with pytest.raises(client_exceptions.MicroserviceManagementClientError) as excinfo:
+                    ret_value = ms_mgt_client.delete_configuration_item("TEST", "blah")
```
Comment (ashish-jabble) [inline tests/unit/python/foglamp/common/microservice_management_client/test_microservice_management_client.py]: same as ^^ and fix standard import order for the test file

File: python/foglamp/services/core/api/configuration.py [hunk 13]
```diff
@@ -95,8 +95,12 @@
         category_desc = data.get('description')
         category_value = data.get('value')
 
+        should_keep_original_items = data.get('keep_original_items', False)
+        if not isinstance(should_keep_original_items, bool):
+            raise web.HTTPBadRequest(reason="keep_original_items should be boolean true | false")
```
Comment (ashish-jabble) [inline python/foglamp/services/core/api/configuration.py]: It would be better, if we add 1 test into test_configuration.py

**Pull request merged**